boa/src/builtins/number/mod.rs

```diff
@@ -555,6 +555,221 @@ impl Number {
+    ///  - [MDN documentation][mdn]
+    ///
+    /// [spec]: https://tc39.es/ecma262/#sec-isnan-number
+    /// [mdn]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number/isNaN
+    pub(crate) fn number_is_nan(
+        _this: &Value,
+        args: &[Value],
+        _ctx: &mut Interpreter,
+    ) -> ResultValue {
+        Ok(Value::Boolean(if let Some(val) = args.get(0) {
+            match val {
+                Value::Integer(_) => false,
+                Value::Rational(number) => number.is_nan(),
+                Value::Object(ref obj) => {
+                    let obj = obj.borrow();
+                    if let Some(number) = obj.as_number() {
+                        number.is_nan()
+                    } else {
+                        false
+                    }
+                }
+                _ => false,
+            }
+        } else {
+            false
+        }))
+    }
+
+    /// `Number.isSafeInteger( number )`
+    ///
+    /// Checks if the argument is an integer, returning false if it isn't.
+    ///
+    /// If abs(number) ≤ MAX_SAFE_INTEGER true is returned.
+    ///
+    /// Otherwise false is returned.
+    ///
+    /// More information:
+    ///  - [ECMAScript reference][spec]
+    ///  - [MDN documentation][mdn]
+    ///
+    /// [spec]: https://tc39.es/ecma262/#sec-isnan-number
+    /// [mdn]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number/isNaN
+    pub(crate) fn is_safe_integer(
+        _this: &Value,
+        args: &[Value],
+        _ctx: &mut Interpreter,
+    ) -> ResultValue {
+        Ok(Value::Boolean(if let Some(val) = args.get(0) {
+            if let Some(integer) = Number::is_integer(val) {
+                integer.abs() <= Number::MAX_SAFE_INTEGER
+            } else {
+                false
+            }
+        } else {
+            false
+        }))
+    }
+
+    /// Checks if the argument is a finite integer Number value.
+    ///
+    /// More information:
+    ///  - [ECMAScript reference][spec]
+    ///
+    /// [spec]: https://tc39.es/ecma262/#sec-isinteger
+    #[allow(clippy::float_cmp)]
+    pub(crate) fn is_integer(val: &Value) -> Option<f64> {
+        match val {
+            Value::Integer(number) => Some((*number).into()),
+            Value::Rational(number) => {
+                if number.is_finite() && number.abs().floor() == number.abs() {
+                    Some(*number)
+                } else {
+                    None
+                }
+            }
+            Value::Object(obj) => {
+                let obj = obj.borrow();
+                if let Some(number) = obj.as_number() {
+                    if number.abs().floor() == number.abs() {
+                        Some(number)
+                    } else {
+                        None
+                    }
+                } else {
+                    None
+                }
+            }
+            _ => None,
+        }
+    }
```

Comment:
```suggestion
    pub(crate) fn is_integer(val: &Value) -> bool {
        match val {
            Value::Integer(number) => true,
            Value::Rational(number) => {
                number.is_finite() && number.abs().floor() == number.abs()
            }
            _ => false,
        }
    }
```
`isInteger` does not check for number object <https://tc39.es/ecma262/#sec-isinteger>, also it it returns a `bool` I don't see the reason for `Option<f64>` could you explain this.

Reply:
The reason I wrote it like that is that the is_safe_integer() method then uses the value as a float to compare to MAX_SAFE_INTEGER, so it would save matching the value again.

Reply:
> The reason I wrote it like that is that the is_safe_integer() method then uses the value as a float to compare to MAX_SAFE_INTEGER, so it would save matching the value again.

Yes. but it would be better if do something like:
```rust
#[inline]
pub(crate) fn is_integer(number: f64) -> bool {
	number.is_finite() && number.abs().floor() == number.abs()
}
```

and in the `is_safe_integer()`:
```rust
	pub(crate) fn is_safe_integer(
        _this: &Value,
        args: &[Value],
        _ctx: &mut Interpreter,
    ) -> ResultValue {
        match args.get(0) {
			Some(Value::Integer(_)) => true,
			Some(Value::Number(number)) if Self::is_integer(number) => {
				number.abs() <= Number::MAX_SAFE_INTEGER
			}
			_ => false 
		}
    }
```

The other way was slightly more inefficient because when it was an `Value::Integer` (is is always a safe integer) we only need to return true, instead of conversion it to `f64` and than doing:
```
if Self::is_integer(number) => {
	number.abs() <= Number::MAX_SAFE_INTEGER
}
```
this to check if the already safe integer is a safe integer. also it removes the check from having an `Option`